src/paho/mqtt/client.py

```diff
@@ -2249,18 +2249,26 @@ def _messages_reconnect_reset_out(self):
                         m.state = mqtt_ms_publish
                     elif m.qos == 2:
                         # self._inflight_messages = self._inflight_messages + 1
-                        if m.state == mqtt_ms_wait_for_pubcomp:
-                            m.state = mqtt_ms_resend_pubrel
-                            m.dup = True
-                        else:
-                            if m.state == mqtt_ms_wait_for_pubrec:
+                        if self._clean_session is True:
+                            if m.state != mqtt_ms_publish:
                                 m.dup = True
```

Comment:
As said in PR global comment, with clean_session=True we should discard any message for which the publish packet was send.

Reply:
If I drop the "is True" some test cases(03-publish-c2b-qos1-disconnect and maybe more) fail. Why are you passing clean_session=-1 in test cases?
````
run = -1`
mqttc = mqtt.Client("publish-qos1-test", run)
````

Should I update the _messages_reconnect_reset_out to clear the _out_messages when clean_session=True and what should I do about testcases?

Reply:
It's expected that this test fail :)
I don't really know why there is this "run = -1" at all... in all test I looked that I looked for it's a constant used in a "while run == -1" (which is thus equivalent to while True).
In this test is used to set clean_session=-1 (equivalent to clean_session=True)... which make the test wrong since this test expect a session to persist a disconnection or timeout.

Reply:
Ok cool! I see you have updated the test cases. I have modified the pull request to completely remove _out_messages messages when clean_session=True.